File: backend/scoring.py
"""Answer key comparison, scoring, and result aggregation."""
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class QuestionResult:
    question: int
    correct_answer: Optional[int]
    student_answer: Optional[int]
    status: str          # "correct" | "wrong" | "blank" | "multiple" | "not_detected"
    marks: float


@dataclass
class StudentResult:
    student_id: str
    student_name: str
    total_questions: int
    correct: int = 0
    wrong: int = 0
    blank: int = 0
    multiple: int = 0
    not_detected: int = 0
    total_marks: float = 0.0
    max_marks: float = 0.0
    percentage: float = 0.0
    question_results: list = field(default_factory=list)
    quality_warnings: list = field(default_factory=list)


def score_sheet(scan_answers: dict, answer_key: dict,
                 marks_per_correct: float = 1.0,
                 negative_marking: float = 0.0,
                 student_id: str = "", student_name: str = "") -> StudentResult:
    """
    scan_answers: dict[int question] -> BubbleResult (from omr_engine)
    answer_key:   dict[int question] -> int correct_option (1-4)

    The sheet is scored over the full set of questions present (the union of
    the answer key and the questions detected on the scanned sheet). This keeps
    the maximum marks tied to the real number of questions on the sheet (e.g.
    40) even if a correct answer for one question was never set in the key,
    instead of silently shrinking the total to the number of saved key entries.
    """
    all_questions = set(answer_key.keys()) | set(scan_answers.keys())
    total_q = max(all_questions) if all_questions else 0
    result = StudentResult(
        student_id=student_id,
        student_name=student_name,
        total_questions=total_q,
        max_marks=total_q * marks_per_correct,
    )

    for q in range(1, total_q + 1):
        correct_opt = answer_key.get(q)
        bubble = scan_answers.get(q)

        if correct_opt is None:
            # No correct answer configured for this question: it still counts
            # toward the total (so the sheet stays out of its full size), but
            # it cannot be graded, so it is treated as blank (0 marks).
            status, marks = "blank", 0.0
            result.blank += 1
        elif bubble is None or (bubble.flagged and bubble.flag_reason == "not detected"):
            status, marks = "not_detected", 0.0
            result.not_detected += 1
        elif bubble.is_multiple:
            status, marks = "multiple", -negative_marking
            result.multiple += 1
        elif bubble.selected is None:
            status, marks = "blank", 0.0
            result.blank += 1
        elif bubble.selected == correct_opt:
            status, marks = "correct", marks_per_correct
            result.correct += 1
        else:
            status, marks = "wrong", -negative_marking
            result.wrong += 1

        result.total_marks += marks
        result.question_results.append(QuestionResult(
            question=q,
            correct_answer=correct_opt,
            student_answer=(bubble.selected if bubble else None),
            status=status,
            marks=marks,
        ))

    result.total_marks = max(0.0, round(result.total_marks, 2))
    result.percentage = round(
        (result.total_marks / result.max_marks * 100) if result.max_marks else 0.0, 2
    )
    return result


# ---------------------------------------------------------------------------
# NMMS subject split. A single scanned sheet holds 40 answers; these fixed
# question ranges group them into the four NMMS निकालपत्रक subjects. एकूण
# (total) is the sum of the four subject marks. Ranges are inclusive and
# 1-based; adjust here if the real sheet layout differs.
# ---------------------------------------------------------------------------
NMMS_SUBJECT_RANGES = {
    "बुद्धिमत्ता": (1, 10),
    "विज्ञान": (11, 20),
    "स.शास्त्र": (21, 30),
    "गणित": (31, 40),
}


def subject_breakdown(result: StudentResult, ranges: dict | None = None) -> dict:
    """Split a scored sheet into per-subject marks + एकूण total.

    Counts a question toward its subject only when it was answered correctly,
    using the same question numbering as `question_results`. Returns
    ``{"subjects": {name: marks}, "total": marks}`` so the four cells + एकूण
    can be filled straight into the निकालपत्रक Excel from a single scan.
    """
    ranges = ranges or NMMS_SUBJECT_RANGES
    correct_by_q = {
        qr.question for qr in result.question_results if qr.status == "correct"
    }
    subjects = {}
    total = 0
    for name, (lo, hi) in ranges.items():
        marks = sum(1 for q in range(lo, hi + 1) if q in correct_by_q)
        subjects[name] = marks
        total += marks
    return {"subjects": subjects, "total": total}


def grade_for_percentage(pct: float) -> str:
    if pct >= 90: return "A+"
    if pct >= 80: return "A"
    if pct >= 70: return "B+"
    if pct >= 60: return "B"
    if pct >= 50: return "C"
    if pct >= 40: return "D"
    return "F"


def batch_dashboard(results: list[StudentResult]) -> dict:
    if not results:
        return {}
    marks = [r.total_marks for r in results]
    pass_mark_pct = 40
    passed = sum(1 for r in results if r.percentage >= pass_mark_pct)
    # per-question difficulty: % of students who got it wrong
    q_wrong_count = {}
    q_total = {}
    for r in results:
        for qr in r.question_results:
            q_total[qr.question] = q_total.get(qr.question, 0) + 1
            if qr.status in ("wrong", "blank", "multiple"):
                q_wrong_count[qr.question] = q_wrong_count.get(qr.question, 0) + 1
    difficulty = sorted(
        [(q, round(q_wrong_count.get(q, 0) / q_total[q] * 100, 1)) for q in q_total],
        key=lambda x: -x[1]
    )
    return {
        "students_scanned": len(results),
        "highest_marks": max(marks),
        "lowest_marks": min(marks),
        "average_marks": round(sum(marks) / len(marks), 2),
        "pass_percent": round(passed / len(results) * 100, 1),
        "fail_percent": round(100 - passed / len(results) * 100, 1),
        "hardest_questions": difficulty[:10],
    }

File: android/app/src/main/java/com/omr/capture/Scoring.kt
package com.omr.capture

/** On-device port of `scoring.py`: compares detected answers with a key. */
object Scoring {

    data class QuestionResult(
        val question: Int,
        val correctAnswer: Int?,
        val studentAnswer: Int?,
        val status: String,   // correct | wrong | blank | multiple | not_detected
        val marks: Double,
    )

    data class StudentResult(
        val totalQuestions: Int,
        var correct: Int = 0,
        var wrong: Int = 0,
        var blank: Int = 0,
        var multiple: Int = 0,
        var notDetected: Int = 0,
        var totalMarks: Double = 0.0,
        var maxMarks: Double = 0.0,
        var percentage: Double = 0.0,
        val questionResults: MutableList<QuestionResult> = mutableListOf(),
    )

    fun score(
        scanAnswers: Map<Int, OmrEngine.BubbleResult>,
        answerKey: Map<Int, Int>,
        marksPerCorrect: Double = 1.0,
        negativeMarking: Double = 0.0,
    ): StudentResult {
        val result = StudentResult(
            totalQuestions = answerKey.size,
            maxMarks = answerKey.size * marksPerCorrect,
        )

        for (q in answerKey.keys.sorted()) {
            val correctOpt = answerKey[q]
            val bubble = scanAnswers[q]

            val (status, marks) = when {
                bubble == null || (bubble.flagged && bubble.flagReason == "not detected") -> {
                    result.notDetected++; Pair("not_detected", 0.0)
                }
                bubble.isMultiple -> { result.multiple++; Pair("multiple", -negativeMarking) }
                bubble.selected == null -> { result.blank++; Pair("blank", 0.0) }
                bubble.selected == correctOpt -> { result.correct++; Pair("correct", marksPerCorrect) }
                else -> { result.wrong++; Pair("wrong", -negativeMarking) }
            }

            result.totalMarks += marks
            result.questionResults.add(
                QuestionResult(q, correctOpt, bubble?.selected, status, marks)
            )
        }

        result.totalMarks = maxOf(0.0, round2(result.totalMarks))
        result.percentage = round2(
            if (result.maxMarks > 0) result.totalMarks / result.maxMarks * 100 else 0.0
        )
        return result
    }

    fun grade(pct: Double): String = when {
        pct >= 90 -> "A+"; pct >= 80 -> "A"; pct >= 70 -> "B+"
        pct >= 60 -> "B"; pct >= 50 -> "C"; pct >= 40 -> "D"; else -> "F"
    }

    private fun round2(v: Double) = Math.round(v * 100.0) / 100.0
}
